demos/https/iot_demo_https_s3_download_async.c

```diff
@@ -170,23 +170,109 @@
+
+/**
+ * @brief Bitmask calculation for setting the file download information bitmaps.
+ */
+#define BITMASK( rangeBlock )        ( ( ( uint32_t ) 1 ) << ( ( ( uint32_t ) rangeBlock ) % BITS_PER_BYTE ) )
+
+/**
+ * @brief Byte offset calculation for indexing to the correct byte in the file download information bitmap.
+ */
+#define BYTE_OFFSET( rangeBlock )    ( ( ( uint32_t ) rangeBlock ) >> LOG2_BITS_PER_BYTE )
 
 /*-----------------------------------------------------------*/
 
 /**
  * @brief This structure defines data consumed and kept track of in the async callbacks per request.
  */
-typedef struct _asyncDownloadData
+typedef struct _requestData
 {
-    char rangeValueStr[ RANGE_VALUE_MAX_LENGTH ]; /* This string is generated outside of the callback context and used
-                                                   * to set the request Range header within the callback context. */
-    int reqNum;                                   /* This is the current request number in the static pool of request memory. */
-    uint32_t currDownloaded;                      /* The current number of bytes downloaded in this request. */
-    uint32_t numReqBytes;                         /* The number of bytes we want to downloaded in this request. */
-} _asyncDownloadData_t;
+    char rangeValueStr[ RANGE_VALUE_MAX_LENGTH ]; /**< @brief This string is generated outside of the callback context and used to set the request Range header within the callback context. */
+    int reqNum;                                   /**< @brief This is the current request number in the static pool of request memory. */
+    uint32_t currRange;                           /**< @brief This is the current range block that is being downloaded. */
+    uint32_t currDownloaded;                      /**< @brief The current number of bytes downloaded in this request. */
+    uint32_t numReqBytes;                         /**< @brief The number of bytes we want to downloaded in this request. */
+    IotHttpsConnectionHandle_t * pConnHandle;     /**< @brief The connection handle for the request/response. */
+    IotHttpsConnectionInfo_t * pConnConfig;       /**< @brief Pointer to the connection handle for the request/response. */
+    bool scheduled;                               /**< @brief This is set to true when the request has been scheduled */
+} _requestData_t;
+
+/**
+ * @brief This structure defines information related to the file we wish to download.
+ *
+ * This information keeps track of the size of the file, the total number of range blocks, which blocks have completed,
+ * which ones are scheduled to be requested, and how many range blocks remain.
+ */
+typedef struct _fileDownloadInfo
+{
+    uint32_t fileSize;          /**< @brief The total size of the file to be downloaded. */
+    int32_t totalRanges;        /**< @brief the total number of ranges in the file size. */
```

Comment:
For comparison with rangesRemaining.

Reply:
Maybe nobody will use our demo to download Gigabyte file, but probably still better to document it only works with file size less than 2^31 bytes.

Reply:
Sure

Reply:
> This is the length of the downloadedBitmap array? If so, could just call it "...Length".
> 
> Why is it signed?

I feel it is cleared to readers to note it as all the possible range blocks instead of the bitmap length.